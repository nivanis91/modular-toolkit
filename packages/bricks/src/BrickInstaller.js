import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class BrickInstaller extends Component {
    constructor(props) {
        super(props);
        const { brickManager, bricks } = props;
        brickManager.installBricks(bricks);
    }

    render() {
        const { WrappedComponent, wrappedComponentProps } = this.props;
        return <WrappedComponent {...wrappedComponentProps} />;
    }
}

BrickInstaller.propTypes = {
    brickManager: PropTypes.object.isRequired,
    bricks: PropTypes.object.isRequired,
    WrappedComponent: PropTypes.func.isRequired,
    wrappedComponentProps: PropTypes.object
};
