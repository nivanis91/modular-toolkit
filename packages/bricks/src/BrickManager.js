import { addValueByDottedPath, getValueByDottedPath, mergeReducers } from './utils';
import { registerSelectorsForUseWithGlobalState } from '@modular-toolkit/selectors';

/* exported for testing only */
export const replaceReducer = Symbol(
    process.env.NODE_ENV === 'production' ? undefined : 'replaceReducer (private method of BrickManager)'
);

const addSelectors = Symbol(
    process.env.NODE_ENV === 'production' ? undefined : 'addSelectors (private method of BrickManager)'
);

const addSaga = Symbol(process.env.NODE_ENV === 'production' ? undefined : 'addSaga (private method of BrickManager)');

const saveReducer = Symbol(
    process.env.NODE_ENV === 'production' ? undefined : 'saveReducer (private method of BrickManager)'
);

const loadReducer = Symbol(
    process.env.NODE_ENV === 'production' ? undefined : 'loadReducer (private method of BrickManager)'
);

const hasReducer = Symbol(
    process.env.NODE_ENV === 'production' ? undefined : 'hasReducer (private method of BrickManager)'
);

export default class {
    constructor({ store, reducer = s => s, sagaMiddleware }) {
        this.store = store;
        this.originalReducer = reducer;
        this.additionalReducers = {};
        this.sagaMiddleware = sagaMiddleware;
    }

    installBrick(storePath, brick) {
        this.installBricks({ [storePath]: brick });
    }

    installBricks(bricks) {
        let changed = false;
        Object.entries(bricks)
            .filter(([storePath]) => !this[hasReducer](storePath))
            .forEach(([storePath, { reducer, selectors, saga }]) => {
                changed = true;
                this[saveReducer](storePath, reducer);
                this[addSelectors](storePath, selectors);
                this[addSaga](saga);
            });
        if (changed) {
            this[replaceReducer]();
        }
    }

    [replaceReducer]() {
        this.store.replaceReducer(mergeReducers(this.originalReducer, this.additionalReducers));
    }

    [saveReducer](storePath, reducer) {
        this.additionalReducers = addValueByDottedPath(this.additionalReducers, storePath, reducer);
    }

    [addSelectors](storePath, selectors) {
        registerSelectorsForUseWithGlobalState(storePath, selectors);
    }

    [addSaga](saga) {
        this.sagaMiddleware.run(saga);
    }

    [loadReducer](storePath) {
        return getValueByDottedPath(this.additionalReducers, storePath);
    }

    [hasReducer](storePath) {
        return this[loadReducer](storePath) !== null;
    }
}
