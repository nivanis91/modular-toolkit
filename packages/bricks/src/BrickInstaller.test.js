import React from 'react';
import BrickInstaller from './BrickInstaller';
import { mount } from 'enzyme';
import BrickManager from './BrickManager';
import PropTypes from 'prop-types';

function MyComponent({ name }) {
    return <h1>Hello {name}!</h1>;
}

MyComponent.propTypes = {
    name: PropTypes.string.isRequired
};

jest.mock('./BrickManager');

const brickManager = new BrickManager();
const bricks = Symbol('bricks');

describe('When I mount a component that is wrapped with a BrickInstaller', () => {
    let component;
    beforeEach(() =>
        (component = mount(
            <BrickInstaller
                brickManager={brickManager}
                bricks={bricks}
                WrappedComponent={MyComponent}
                wrappedComponentProps={{
                    name: 'world'
                }}
            />
        )));
    describe("the provided brick manager's “installBricks” method", () =>
        it('is called with the provided brick map', () => expect(brickManager.installBricks).toBeCalledWith(bricks)));
    describe('the provided wrapped component', () => {
        it('is rendered with the provided wrapped component props', () => {
            expect(component.html()).toMatchSnapshot();
        });
    });
});
