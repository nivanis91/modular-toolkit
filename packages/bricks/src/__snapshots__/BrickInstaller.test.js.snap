// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I mount a component that is wrapped with a BrickInstaller the provided wrapped component is rendered with the provided wrapped component props 1`] = `"<h1>Hello world!</h1>"`;
